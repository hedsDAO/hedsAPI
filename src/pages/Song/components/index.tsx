export { default as SongDetails } from './SongDetails';
export { default as SongHeader } from './SongHeader';
export { default as SongWaveform } from './SongWaveform';
export { default as SongAppearsOn } from './SongAppearsOn';
export { default as SongLikes } from './SongLikes';
export { default as SongInsights } from './SongInsights';
