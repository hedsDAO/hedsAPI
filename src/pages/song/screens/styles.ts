import { BoxProps, DividerProps, SkeletonProps } from '@chakra-ui/react';

export const $navSkeletonStyles = (navbarTabs: string[], isLoading: boolean): SkeletonProps => ({
  isLoaded: !!navbarTabs?.length && !isLoading,
  startColor: 'heds.bg3',
  endColor: 'heds.bg1',
  w: '100vw',
});

export const $navBoxStyles: BoxProps = {
  maxW: { lg: '85vw' },
  mx: 'auto',
};

export const $dividerStyles: DividerProps = {
  mt: { lg: 2 },
  borderColor: 'heds.100',
};

export const $contentBoxStyles: BoxProps = {
  px: 6,
  maxW: { lg: '8xl' },
  mx: 'auto',
};
