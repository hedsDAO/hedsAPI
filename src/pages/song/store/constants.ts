
export const testData = {
    track: 'internal dingo',
    duration: 59,
    type: 1,
    space: 'heds',
    stats: {
      likedBy: {},
      plays: 1,
      likes: 0,
    },
    artist: 'harris cole',
    cover:
      'https://firebasestorage.googleapis.com/v0/b/heds-104d8.appspot.com/o/covers%2Fhedstape%2F2.png?alt=media&token=b37ac39b-243e-411f-820f-17c9b8e454df',
    no: 4,
    id: '2',
    tape: 'hedstape',
    subImage: 'https://www.heds.cloud/ipfs/Qmdyz7xSWZjDBBe9uRS8QU4NFRCP24zjExVJLS4mbkTshv',
    album: 'hedsTAPE 02',
    public: true,
    wallet: '0x8e09b494f20aeadb8d9b3c8b3bec7cb6c7e18e83',
    subId: 'internalDINGO',
    audio: 'https://www.heds.cloud/ipfs/QmNMrUBhK5wUwjVDhADvMQnh4gq3d2hhZjKwu6rXjonaHr',
  };