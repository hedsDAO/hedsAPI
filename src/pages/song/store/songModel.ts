import { createModel } from '@rematch/core';
import type { RootModel } from '@/models';

interface Contributor {
  userId: string;
  split: number;
  verified: boolean;
}

interface SubmissionData {
  subId: string;
  subImage: string;
  proposalId: string;
}

interface TrackStats {
  likedBy: { [key: string]: number };
  plays: number;
}

export enum TrackType {
  SUBMISSION,
  TRACK,
  COLLAB,
  SAMPLE,
  HEDSOLO,
}

export interface SongState {
  audio: string;
  contributors: Contributor[];
  cover: string;
  duration: number;
  public: boolean;
  stats: TrackStats;
  appearsOn: string[];
  trackName: string;
  type: TrackType;
  submissionData: SubmissionData;
  cyaniteId: string;
  created: number;
}

export const songModel = createModel<RootModel>()({
  state: {
    audio: '',
    contributors: [],
    cover: 'string',
    duration: 0,
    public: true,
    stats: {},
    appearsOn: [],
    trackName: '',
    type: 0,
    submissionData: {},
    cyaniteId: '',
    created: 0,
  } as SongState,
});
