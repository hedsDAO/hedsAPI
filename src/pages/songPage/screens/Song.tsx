export const Song = () => <>song page</>;
