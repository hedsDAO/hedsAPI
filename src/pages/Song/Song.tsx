import { useEffect, useState } from 'react';
import { Box } from '@chakra-ui/react';
import { useSelector } from 'react-redux';
import { store } from '@/store';
import { mockArtist, mockSong, mockTape } from './models/constant';
import { SongAppearsOn, SongDetails, SongHeader, SongWaveform } from './components';
import { Nav } from '@/components/Nav/Nav';
import { SongNavbarTabs } from './models/common';
import { getCyaniteData } from '@/utils';

export const Song = () => {
  const [cyaniteData, setCyaniteData] = useState(null);
  const currentTab = useSelector(store.select.navModel.selectCurrentTab);
  useEffect(() => {
    const fetchCyaniteData = async () => {
      let data = await getCyaniteData(parseInt(mockSong.cyanite_id));
      setCyaniteData(data.result);
    };
    fetchCyaniteData();
  }, []);
  return (
    <Box>
      <SongHeader />
      <SongWaveform />
      <Nav tabs={['Details', 'Appears On', 'Likes']} />
      {currentTab === SongNavbarTabs.DETAILS && <SongDetails cyanite={cyaniteData} curator={mockArtist} tape={mockTape} />}
      {currentTab === SongNavbarTabs.APPEARS_ON && <SongAppearsOn />}
      {currentTab === SongNavbarTabs.LIKES && <>Likes</>}
    </Box>
  );
};
