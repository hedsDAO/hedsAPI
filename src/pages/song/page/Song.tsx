import { useEffect, useRef, useState } from 'react';
import { formWaveSurferOptions } from '@/utils';
import { useDispatch, useSelector } from 'react-redux';
import { Dispatch, RootState } from '@/store';
import WaveSurfer from 'wavesurfer.js';
import { WaveSurferParams } from 'wavesurfer.js/types/params';

// Components
import { AspectRatio, Flex, Grid, GridItem, Image, Stack, Tabs, TabList, TabPanels, Tab, TabPanel, Text } from '@chakra-ui/react';
import { TapeDataTab } from '../components/TapeDataTab';
import { AppearsOnTab } from '../components/AppearsOnTab';
import { LikesTab } from '../components/LikesTab';
import { PlayIcon } from '@heroicons/react/24/solid';
import { HeartIcon } from '@heroicons/react/24/outline';

const testData = {
  track: 'internal dingo',
  duration: 59,
  type: 0,
  space: 'heds',
  stats: {
    likedBy: {},
    plays: 1,
    likes: 0,
  },
  artist: 'harris cole',
  cover: 'https://firebasestorage.googleapis.com/v0/b/heds-104d8.appspot.com/o/covers%2Fhedstape%2F2.png?alt=media&token=b37ac39b-243e-411f-820f-17c9b8e454df',
  no: 4,
  id: '2',
  tape: 'hedstape',
  subImage: 'https://www.heds.cloud/ipfs/Qmdyz7xSWZjDBBe9uRS8QU4NFRCP24zjExVJLS4mbkTshv',
  album: 'hedsTAPE 02',
  public: true,
  wallet: '0x8e09b494f20aeadb8d9b3c8b3bec7cb6c7e18e83',
  subId: 'internalDINGO',
  audio: 'https://www.heds.cloud/ipfs/QmNMrUBhK5wUwjVDhADvMQnh4gq3d2hhZjKwu6rXjonaHr',
  created: 1679363005000,
};

export const Song = () => {
  const dispatch = useDispatch<Dispatch>();
  const [isPlaying, setIsPlaying] = useState(false);
  const waveformRef = useRef<HTMLDivElement | null>(null);
  const wavesurfer = useRef<WaveSurfer | null>();

  useEffect(() => {
    if (testData.audio?.length) {
      dispatch.voteModel.setIsLoading(true);
      var options; // wavesurfer params
      if (waveformRef.current) {
        options = {
          container: waveformRef.current,
          waveColor: '#000000',
          progressColor: '#C025D3',
          cursorColor: 'transparent',
          barWidth: 2.5,
          barRadius: 0,
          responsive: true,
          height: 220,
          normalize: true,
          hideScrollbar: true,
        };
      }
      if (options) wavesurfer.current = WaveSurfer.create(options);
      if (typeof testData.audio === 'string') wavesurfer?.current?.load(testData.audio);
      wavesurfer?.current?.on('ready', () => dispatch.voteModel.setIsLoading(false));
      wavesurfer?.current?.on('finish', () => {
        setIsPlaying(false);
        wavesurfer?.current?.stop();
      });
    }
    return () => {
      setIsPlaying(false);
      if (wavesurfer?.current) {
        wavesurfer?.current?.destroy();
      }
    };
  }, []);

  return (
    <>
      <Grid maxW="full" templateColumns="repeat(12, 1fr)" gap={6}>
        <GridItem colSpan={7}>
          <Stack mt="-10" pl={20} gap={4} justifyContent={'center'} h="full">
            <Flex direction="column">
              <Text letterSpacing={'widest'} fontSize={'md'} textColor={'gray.500'}>
                TRACK
              </Text>
              <Text letterSpacing={'widest'} fontSize={'3xl'}>
                {testData?.subId}
              </Text>
            </Flex>
            <Flex direction="column">
              <Text letterSpacing={'widest'} fontSize={'md'} textColor={'gray.500'}>
                ARTIST
              </Text>
              <Text letterSpacing={'widest'} fontSize={'3xl'}>
                {testData?.artist}
              </Text>
            </Flex>
            <Flex pt={6} gap={3}>
              <PlayIcon
                width={30}
                onClick={() => {
                  wavesurfer?.current?.playPause();
                  setIsPlaying(!isPlaying);
                }}
              ></PlayIcon>
              <HeartIcon width={30}></HeartIcon>
            </Flex>
          </Stack>
        </GridItem>
        <GridItem colSpan={5}>
          <AspectRatio ratio={1}>
            <Image src={testData.cover} />
          </AspectRatio>
        </GridItem>
      </Grid>
      <Grid gap={2} w="full" alignItems={'center'} templateColumns="repeat(12, 1fr)">
        <GridItem h="110px" overflowY={'hidden'} mt={'-110px'} colSpan={12}>
          <div ref={waveformRef} className={'w-full'}></div>
        </GridItem>
      </Grid>
      <Tabs variant="soft-rounded" colorScheme="blackAlpha" bgColor="blackAlpha.400" p="2rem">
        <TabList>
          <Tab fontSize="sm">TAPE DATA</Tab>
          <Tab fontSize="sm">APPEARS ON</Tab>
          <Tab fontSize="sm">LIKES</Tab>
        </TabList>
        <TabPanels>
          <TabPanel>
            <TapeDataTab
              artist={testData.artist}
              wallet={testData.wallet}
              subId={testData.subId}
              subImage={testData.subImage}
              type={testData.type}
              created={testData.created}
            />
          </TabPanel>
          <TabPanel>
            <AppearsOnTab />
          </TabPanel>
          <TabPanel>
            <LikesTab />
          </TabPanel>
        </TabPanels>
      </Tabs>
    </>
  );
};
