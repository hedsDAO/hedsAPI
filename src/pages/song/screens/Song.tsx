import { useEffect, useRef } from 'react';
import { Box, Divider, Skeleton, keyframes } from '@chakra-ui/react';
import { useDispatch, useSelector } from 'react-redux';
import { Dispatch, store } from '@/store';
import { Header } from '@/pages/song/components/Header/Header';
import { useParams } from 'react-router-dom';
import { useLocalAudio } from '@/pages/song/components/useLocalAudio/useLocalAudio';
import { Waveform } from '@/pages/song/components/Waveform/Waveform';
import { Nav } from '@/components/Nav/Nav';
import { SongNavbarTabs } from '@/pages/song/models/common';
import { Details } from '@/pages/song/components/Details/Details';
import { AppearsOn } from '@/pages/song/components/AppearsOn/AppearsOn';
import { Related } from '../components/Related/Related';
import { Likes } from '@/pages/song/components/Likes/Likes';

export const Song = () => {
  const dispatch = useDispatch<Dispatch>();
  const { id } = useParams();
  const waveformRef = useRef<HTMLDivElement>(null);
  const { handlePlayPause } = useLocalAudio(waveformRef);
  const song = useSelector(store.select.songModel.selectSong);
  const navbarTabs = useSelector(store.select.songModel.selectNavbarTabs);
  const currentTab = useSelector(store.select.navModel.selectCurrentTab);
  const isLoading = useSelector(store.select.songModel.selectIsLoading);


  useEffect(() => {
    dispatch.navModel.setCurrentTab(SongNavbarTabs.DETAILS);
    if (id?.length) dispatch.songModel.getSongData(id);
  }, [id]);

  return (
    <Box>
      <Header handlePlayPause={handlePlayPause} />
      {song?.audio && <Waveform waveformRef={waveformRef} />}
      <Skeleton isLoaded={!!navbarTabs?.length && !isLoading} startColor="heds.bg3" endColor="heds.bg1" w="100vw" fitContent>
        <Box maxW={{ lg: '85vw' }} mx="auto">
          <Nav tabs={navbarTabs || ['Details', 'Related']} />
        </Box>
      </Skeleton>
      <Divider mt={{ lg: 2 }} borderColor="heds.100" />
      <Box px={6} maxW={{ lg: '8xl' }} mx="auto">
        {currentTab === SongNavbarTabs.DETAILS && <Details />}
        {currentTab === SongNavbarTabs.RELATED && <Related />}
        {currentTab === SongNavbarTabs.LIKES && <Likes />}
        {currentTab === SongNavbarTabs.APPEARS_ON && <AppearsOn />}
      </Box>
    </Box>
  );
};
