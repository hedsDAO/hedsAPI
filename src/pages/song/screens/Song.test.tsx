import { BrowserRouter as Router } from 'react-router-dom';
import { Song } from '@/pages/song/screens/Song';
import { store } from '@/store';
import { renderWithRematchStore } from '@/tests/renderWithRematchStore';

describe('Song unit', () => {
  beforeAll(async () => {});
  describe('snapshot', () => {
    test('snapshot', async () => {
      const snapshot = renderWithRematchStore(
        <Router>
          <Song />
        </Router>,
        store,
      );
      expect(snapshot).toMatchSnapshot();
    });
  });
});
