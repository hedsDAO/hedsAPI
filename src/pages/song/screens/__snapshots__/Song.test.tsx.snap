// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Song unit snapshot snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="css-0"
      >
        <div
          class="css-79elbk"
        >
          <div
            class="css-y7fuh1"
          >
            <div
              class="chakra-skeleton css-i7le5z"
            >
              <div
                class="chakra-aspect-ratio css-rioh04"
              >
                <img
                  class="chakra-image css-rdbdr7"
                />
              </div>
            </div>
            <div
              class="css-1prtw9i"
            >
              <div
                class="chakra-skeleton css-10pngdb"
              >
                <span
                  class="chakra-avatar css-yie7qj"
                >
                  <svg
                    aria-label=" avatar"
                    class="chakra-avatar__svg css-16ite8i"
                    role="img"
                    viewBox="0 0 128 128"
                  >
                    <path
                      d="M103,102.1388 C93.094,111.92 79.3504,118 64.1638,118 C48.8056,118 34.9294,111.768 25,101.7892 L25,95.2 C25,86.8096 31.981,80 40.6,80 L87.4,80 C96.019,80 103,86.8096 103,95.2 L103,102.1388 Z"
                      fill="currentColor"
                    />
                    <path
                      d="M63.9961647,24 C51.2938136,24 41,34.2938136 41,46.9961647 C41,59.7061864 51.2938136,70 63.9961647,70 C76.6985159,70 87,59.7061864 87,46.9961647 C87,34.2938136 76.6985159,24 63.9961647,24"
                      fill="currentColor"
                    />
                  </svg>
                </span>
              </div>
            </div>
          </div>
          <div
            class="css-13ix392"
          >
            <div
              class="css-165ru3g"
            >
              <button
                class="chakra-button css-269m92"
                data-testid="song-play-pause-button"
                type="button"
              >
                <i
                  class="chakra-text fas fa-play css-6kfj2u"
                />
              </button>
              <div
                class="chakra-stack css-c0f9yd"
              >
                <div
                  class="chakra-stack css-33dobs"
                >
                  <div
                    class="css-dxxn3y"
                  >
                    <p
                      class="chakra-text css-90p3gp"
                    >
                      ARTIST
                    </p>
                  </div>
                  <hr
                    aria-orientation="horizontal"
                    class="chakra-divider css-1flx70t"
                  />
                </div>
                <p
                  class="chakra-text css-1vvpe8o"
                />
              </div>
              <button
                class="chakra-button css-focya0"
                data-testid="song-like-button"
                disabled=""
                type="button"
              >
                <i
                  class="chakra-text fas fa-heart css-e5peyk"
                />
              </button>
            </div>
            <div
              class="css-jmqpxj"
            >
              <div
                class="chakra-aspect-ratio css-rioh04"
              >
                <div
                  class="css-0"
                >
                  <div
                    class="chakra-skeleton css-cdkrf0"
                  >
                    <img
                      class="chakra-image css-1h57lvz"
                    />
                  </div>
                  <span
                    class="chakra-avatar css-yie7qj"
                  >
                    <svg
                      aria-label=" avatar"
                      class="chakra-avatar__svg css-16ite8i"
                      role="img"
                      viewBox="0 0 128 128"
                    >
                      <path
                        d="M103,102.1388 C93.094,111.92 79.3504,118 64.1638,118 C48.8056,118 34.9294,111.768 25,101.7892 L25,95.2 C25,86.8096 31.981,80 40.6,80 L87.4,80 C96.019,80 103,86.8096 103,95.2 L103,102.1388 Z"
                        fill="currentColor"
                      />
                      <path
                        d="M63.9961647,24 C51.2938136,24 41,34.2938136 41,46.9961647 C41,59.7061864 51.2938136,70 63.9961647,70 C76.6985159,70 87,59.7061864 87,46.9961647 C87,34.2938136 76.6985159,24 63.9961647,24"
                        fill="currentColor"
                      />
                    </svg>
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="chakra-skeleton css-1a4pk4h"
        >
          <div
            class="css-k31un0"
          >
            <div
              class="css-1whh555"
            >
              <button
                class="chakra-button css-wtvw0i"
                type="button"
              >
                Details
              </button>
              <button
                class="chakra-button css-1fy1hmj"
                type="button"
              >
                Related
              </button>
              <button
                class="chakra-button css-1fy1hmj"
                type="button"
              >
                Likes
              </button>
            </div>
            <div
              class="css-182nf6q"
            >
              <button
                aria-controls="menu-list-:r1:"
                aria-expanded="false"
                aria-haspopup="menu"
                class="chakra-button chakra-menu__menu-button css-19lqrls"
                id="menu-button-:r1:"
                type="button"
              >
                <span
                  class="css-xl71ch"
                >
                  <svg
                    class="chakra-icon css-e8xa0h"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M 3 5 A 1.0001 1.0001 0 1 0 3 7 L 21 7 A 1.0001 1.0001 0 1 0 21 5 L 3 5 z M 3 11 A 1.0001 1.0001 0 1 0 3 13 L 21 13 A 1.0001 1.0001 0 1 0 21 11 L 3 11 z M 3 17 A 1.0001 1.0001 0 1 0 3 19 L 21 19 A 1.0001 1.0001 0 1 0 21 17 L 3 17 z"
                      fill="currentColor"
                    />
                  </svg>
                </span>
              </button>
              <div
                class="css-0"
                style="visibility: hidden; position: absolute; min-width: max-content; inset: 0 auto auto 0;"
              >
                <div
                  aria-orientation="vertical"
                  class="chakra-menu__menu-list css-miyh87"
                  id="menu-list-:r1:"
                  role="menu"
                  style="transform-origin: var(--popper-transform-origin); opacity: 0; visibility: hidden; transform: scale(0.8) translateZ(0);"
                  tabindex="-1"
                >
                  <button
                    class="chakra-menu__menuitem css-3a3z0g"
                    data-index="0"
                    id="menu-list-:r1:-menuitem-:r2:"
                    role="menuitem"
                    tabindex="-1"
                    type="button"
                  >
                    Details
                  </button>
                  <button
                    class="chakra-menu__menuitem css-1b6qfzf"
                    data-index="1"
                    id="menu-list-:r1:-menuitem-:r3:"
                    role="menuitem"
                    tabindex="-1"
                    type="button"
                  >
                    Related
                  </button>
                  <button
                    class="chakra-menu__menuitem css-1b6qfzf"
                    data-index="2"
                    id="menu-list-:r1:-menuitem-:r4:"
                    role="menuitem"
                    tabindex="-1"
                    type="button"
                  >
                    Likes
                  </button>
                </div>
              </div>
              <p
                class="chakra-text css-1fj456c"
              >
                Details
              </p>
            </div>
          </div>
        </div>
        <hr
          aria-orientation="horizontal"
          class="chakra-divider css-1npdkbo"
        />
        <div
          class="css-17o3l8a"
        >
          <div
            class="css-1fz1525"
            data-testid="details-simple-grid"
          >
            <div
              class="css-9xqkvx"
            >
              <div
                class="chakra-stack css-1x65gi9"
              >
                <p
                  class="chakra-text css-fp6k5g"
                  data-testid="song-recent-heading"
                >
                  RECENT
                </p>
                <div
                  class="chakra-stack css-n21gh5"
                />
              </div>
            </div>
            <div
              class="css-9xqkvx"
            >
              <div
                class="chakra-stack css-1x65gi9"
              >
                <p
                  class="chakra-text css-fp6k5g"
                  data-testid="song-attributes-heading"
                >
                  ATTRIBUTES
                </p>
                <div
                  class="css-h3rjfo"
                  data-testid="attributes-grid"
                >
                  <div
                    class="chakra-skeleton css-19z5udh"
                  >
                    <div
                      class="chakra-stack css-1xnv6wj"
                    >
                      <p
                        class="chakra-text css-c4ww8m"
                        data-testid="attribute-name"
                      >
                        KEY
                      </p>
                      <p
                        class="chakra-text css-183fo1i"
                        data-testid="attribute-description"
                      >
                        Unknown Key
                      </p>
                    </div>
                  </div>
                  <div
                    class="chakra-skeleton css-19z5udh"
                  >
                    <div
                      class="chakra-stack css-1xnv6wj"
                    >
                      <p
                        class="chakra-text css-c4ww8m"
                        data-testid="attribute-name"
                      >
                        TIME SIGNATURE
                      </p>
                      <p
                        class="chakra-text css-183fo1i"
                        data-testid="attribute-description"
                      />
                    </div>
                  </div>
                  <div
                    class="chakra-skeleton css-1usnp47"
                  >
                    <div
                      class="css-d9pshq"
                    />
                  </div>
                  <div
                    class="chakra-skeleton css-1usnp47"
                  >
                    <div
                      class="css-d9pshq"
                    />
                  </div>
                  <div
                    class="chakra-skeleton css-1usnp47"
                  >
                    <div
                      class="css-d9pshq"
                    />
                  </div>
                  <div
                    class="chakra-skeleton css-1usnp47"
                  >
                    <div
                      class="css-d9pshq"
                    />
                  </div>
                  <div
                    class="chakra-skeleton css-1usnp47"
                  >
                    <div
                      class="css-d9pshq"
                    />
                  </div>
                  <div
                    class="chakra-skeleton css-1usnp47"
                  >
                    <div
                      class="css-d9pshq"
                    />
                  </div>
                </div>
              </div>
            </div>
            <div
              class="css-9xqkvx"
            >
              <div
                class="chakra-stack css-1x65gi9"
              >
                <p
                  class="chakra-text css-fp6k5g"
                  data-testid="song-mood-heading"
                >
                  MOOD SPECTROGRAM
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="css-0"
    >
      <div
        class="css-79elbk"
      >
        <div
          class="css-y7fuh1"
        >
          <div
            class="chakra-skeleton css-i7le5z"
          >
            <div
              class="chakra-aspect-ratio css-rioh04"
            >
              <img
                class="chakra-image css-rdbdr7"
              />
            </div>
          </div>
          <div
            class="css-1prtw9i"
          >
            <div
              class="chakra-skeleton css-10pngdb"
            >
              <span
                class="chakra-avatar css-yie7qj"
              >
                <svg
                  aria-label=" avatar"
                  class="chakra-avatar__svg css-16ite8i"
                  role="img"
                  viewBox="0 0 128 128"
                >
                  <path
                    d="M103,102.1388 C93.094,111.92 79.3504,118 64.1638,118 C48.8056,118 34.9294,111.768 25,101.7892 L25,95.2 C25,86.8096 31.981,80 40.6,80 L87.4,80 C96.019,80 103,86.8096 103,95.2 L103,102.1388 Z"
                    fill="currentColor"
                  />
                  <path
                    d="M63.9961647,24 C51.2938136,24 41,34.2938136 41,46.9961647 C41,59.7061864 51.2938136,70 63.9961647,70 C76.6985159,70 87,59.7061864 87,46.9961647 C87,34.2938136 76.6985159,24 63.9961647,24"
                    fill="currentColor"
                  />
                </svg>
              </span>
            </div>
          </div>
        </div>
        <div
          class="css-13ix392"
        >
          <div
            class="css-165ru3g"
          >
            <button
              class="chakra-button css-269m92"
              data-testid="song-play-pause-button"
              type="button"
            >
              <i
                class="chakra-text fas fa-play css-6kfj2u"
              />
            </button>
            <div
              class="chakra-stack css-c0f9yd"
            >
              <div
                class="chakra-stack css-33dobs"
              >
                <div
                  class="css-dxxn3y"
                >
                  <p
                    class="chakra-text css-90p3gp"
                  >
                    ARTIST
                  </p>
                </div>
                <hr
                  aria-orientation="horizontal"
                  class="chakra-divider css-1flx70t"
                />
              </div>
              <p
                class="chakra-text css-1vvpe8o"
              />
            </div>
            <button
              class="chakra-button css-focya0"
              data-testid="song-like-button"
              disabled=""
              type="button"
            >
              <i
                class="chakra-text fas fa-heart css-e5peyk"
              />
            </button>
          </div>
          <div
            class="css-jmqpxj"
          >
            <div
              class="chakra-aspect-ratio css-rioh04"
            >
              <div
                class="css-0"
              >
                <div
                  class="chakra-skeleton css-cdkrf0"
                >
                  <img
                    class="chakra-image css-1h57lvz"
                  />
                </div>
                <span
                  class="chakra-avatar css-yie7qj"
                >
                  <svg
                    aria-label=" avatar"
                    class="chakra-avatar__svg css-16ite8i"
                    role="img"
                    viewBox="0 0 128 128"
                  >
                    <path
                      d="M103,102.1388 C93.094,111.92 79.3504,118 64.1638,118 C48.8056,118 34.9294,111.768 25,101.7892 L25,95.2 C25,86.8096 31.981,80 40.6,80 L87.4,80 C96.019,80 103,86.8096 103,95.2 L103,102.1388 Z"
                      fill="currentColor"
                    />
                    <path
                      d="M63.9961647,24 C51.2938136,24 41,34.2938136 41,46.9961647 C41,59.7061864 51.2938136,70 63.9961647,70 C76.6985159,70 87,59.7061864 87,46.9961647 C87,34.2938136 76.6985159,24 63.9961647,24"
                      fill="currentColor"
                    />
                  </svg>
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="chakra-skeleton css-1a4pk4h"
      >
        <div
          class="css-k31un0"
        >
          <div
            class="css-1whh555"
          >
            <button
              class="chakra-button css-wtvw0i"
              type="button"
            >
              Details
            </button>
            <button
              class="chakra-button css-1fy1hmj"
              type="button"
            >
              Related
            </button>
            <button
              class="chakra-button css-1fy1hmj"
              type="button"
            >
              Likes
            </button>
          </div>
          <div
            class="css-182nf6q"
          >
            <button
              aria-controls="menu-list-:r1:"
              aria-expanded="false"
              aria-haspopup="menu"
              class="chakra-button chakra-menu__menu-button css-19lqrls"
              id="menu-button-:r1:"
              type="button"
            >
              <span
                class="css-xl71ch"
              >
                <svg
                  class="chakra-icon css-e8xa0h"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M 3 5 A 1.0001 1.0001 0 1 0 3 7 L 21 7 A 1.0001 1.0001 0 1 0 21 5 L 3 5 z M 3 11 A 1.0001 1.0001 0 1 0 3 13 L 21 13 A 1.0001 1.0001 0 1 0 21 11 L 3 11 z M 3 17 A 1.0001 1.0001 0 1 0 3 19 L 21 19 A 1.0001 1.0001 0 1 0 21 17 L 3 17 z"
                    fill="currentColor"
                  />
                </svg>
              </span>
            </button>
            <div
              class="css-0"
              style="visibility: hidden; position: absolute; min-width: max-content; inset: 0 auto auto 0;"
            >
              <div
                aria-orientation="vertical"
                class="chakra-menu__menu-list css-miyh87"
                id="menu-list-:r1:"
                role="menu"
                style="transform-origin: var(--popper-transform-origin); opacity: 0; visibility: hidden; transform: scale(0.8) translateZ(0);"
                tabindex="-1"
              >
                <button
                  class="chakra-menu__menuitem css-3a3z0g"
                  data-index="0"
                  id="menu-list-:r1:-menuitem-:r2:"
                  role="menuitem"
                  tabindex="-1"
                  type="button"
                >
                  Details
                </button>
                <button
                  class="chakra-menu__menuitem css-1b6qfzf"
                  data-index="1"
                  id="menu-list-:r1:-menuitem-:r3:"
                  role="menuitem"
                  tabindex="-1"
                  type="button"
                >
                  Related
                </button>
                <button
                  class="chakra-menu__menuitem css-1b6qfzf"
                  data-index="2"
                  id="menu-list-:r1:-menuitem-:r4:"
                  role="menuitem"
                  tabindex="-1"
                  type="button"
                >
                  Likes
                </button>
              </div>
            </div>
            <p
              class="chakra-text css-1fj456c"
            >
              Details
            </p>
          </div>
        </div>
      </div>
      <hr
        aria-orientation="horizontal"
        class="chakra-divider css-1npdkbo"
      />
      <div
        class="css-17o3l8a"
      >
        <div
          class="css-1fz1525"
          data-testid="details-simple-grid"
        >
          <div
            class="css-9xqkvx"
          >
            <div
              class="chakra-stack css-1x65gi9"
            >
              <p
                class="chakra-text css-fp6k5g"
                data-testid="song-recent-heading"
              >
                RECENT
              </p>
              <div
                class="chakra-stack css-n21gh5"
              />
            </div>
          </div>
          <div
            class="css-9xqkvx"
          >
            <div
              class="chakra-stack css-1x65gi9"
            >
              <p
                class="chakra-text css-fp6k5g"
                data-testid="song-attributes-heading"
              >
                ATTRIBUTES
              </p>
              <div
                class="css-h3rjfo"
                data-testid="attributes-grid"
              >
                <div
                  class="chakra-skeleton css-19z5udh"
                >
                  <div
                    class="chakra-stack css-1xnv6wj"
                  >
                    <p
                      class="chakra-text css-c4ww8m"
                      data-testid="attribute-name"
                    >
                      KEY
                    </p>
                    <p
                      class="chakra-text css-183fo1i"
                      data-testid="attribute-description"
                    >
                      Unknown Key
                    </p>
                  </div>
                </div>
                <div
                  class="chakra-skeleton css-19z5udh"
                >
                  <div
                    class="chakra-stack css-1xnv6wj"
                  >
                    <p
                      class="chakra-text css-c4ww8m"
                      data-testid="attribute-name"
                    >
                      TIME SIGNATURE
                    </p>
                    <p
                      class="chakra-text css-183fo1i"
                      data-testid="attribute-description"
                    />
                  </div>
                </div>
                <div
                  class="chakra-skeleton css-1usnp47"
                >
                  <div
                    class="css-d9pshq"
                  />
                </div>
                <div
                  class="chakra-skeleton css-1usnp47"
                >
                  <div
                    class="css-d9pshq"
                  />
                </div>
                <div
                  class="chakra-skeleton css-1usnp47"
                >
                  <div
                    class="css-d9pshq"
                  />
                </div>
                <div
                  class="chakra-skeleton css-1usnp47"
                >
                  <div
                    class="css-d9pshq"
                  />
                </div>
                <div
                  class="chakra-skeleton css-1usnp47"
                >
                  <div
                    class="css-d9pshq"
                  />
                </div>
                <div
                  class="chakra-skeleton css-1usnp47"
                >
                  <div
                    class="css-d9pshq"
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            class="css-9xqkvx"
          >
            <div
              class="chakra-stack css-1x65gi9"
            >
              <p
                class="chakra-text css-fp6k5g"
                data-testid="song-mood-heading"
              >
                MOOD SPECTROGRAM
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
